### PR TITLE
TeamSpeak Channel Switch enhancements

### PR DESCRIPTION
**Pull request by tbeswick96**
**When merged this pull request will:**
- Added full string comparison. Will search for exact name first. Break if match found.
- Added word comparison. Will search for word matches second. Skip char comparison if match found.
- Char comparison (Levenshtein distance) occurs third.
- Bit of cleanup

### CODE REVIEW

**Comment by TheMagnetar** [comment]
I would prefer avoiding the use of the `auto` type. It makes the code not so clear in my opinion. Since this is a matter of personal preference, I will not ask for a change unless we all decide to.

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 1]
```diff
@@ -475,38 +473,49 @@ ACRE_RESULT CTS3Client::moveToPreviousTS3Channel() {
...
+        if (details.at(0) != "") {
+            name = details.at(0);
+        }
         while (*channelList) {
             channelId = *channelList;
             channelList++;
+            char* channelName;
             if (ts3Functions.getChannelVariableAsString(ts3Functions.getCurrentServerConnectionHandlerID(), channelId, CHANNEL_NAME, &channelName) == ERROR_ok) {
-                std::string channelNameString = std::string(channelName);
+                auto channelNameString = std::string(channelName);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: This is one of the examples where I think that `auto` should not be used.

**Comment by jonpas** [comment]
I agree with `auto` usage not being preferable, where it should be one specific type, write out the type. Iterators and other things where it's not as specific or readable, `auto` is good.

File: extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp [hunk 2]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
+    return matches;
+}
+
 unsigned int CTS3Client::levenshteinDistance(const std::string& string1, const std::string& string2) {
     int length1 = string1.size();
-    int length2 = string2.size();
+    const int length2 = string2.size();
 
-    auto columnStart = (decltype(length1))1;
+    const auto columnStart = decltype(length1)(1);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp]: Just declare it as the right type, without using auto.

File: extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp [hunk 3]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
+    const auto column = new decltype(length1)[length1 + 1];
     std::iota(column + columnStart, column + length1 + 1, columnStart);
 
     for (auto x = columnStart; x <= length2; x++) {
         column[0] = x;
         auto lastDiagonal = x - columnStart;
         for (auto y = columnStart; y <= length1; y++) {
-            auto oldDiagonal = column[y];
-            auto possibilities = {
+            const auto oldDiagonal = column[y];
```
Comment (TheMagnetar) [inline extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp]: this is also an integer.

File: extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp [hunk 4]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
                 column[y - 1] + 1,
                 lastDiagonal + (string1[y - 1] == string2[x - 1] ? 0 : 1)
             };
-            column[y] = std::min(possibilities);
+            column[y] = min(possibilities);
             lastDiagonal = oldDiagonal;
         }
     }
-    auto result = column[length1];
+    const auto result = column[length1];
```
Comment (TheMagnetar) [inline extensions/src/ACRE2DistortionTestPlugin/TS3Client.cpp]: same

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 5]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
+    return matches;
+}
+
 unsigned int CTS3Client::levenshteinDistance(const std::string& string1, const std::string& string2) {
     int length1 = string1.size();
-    int length2 = string2.size();
+    const int length2 = string2.size();
 
-    auto columnStart = (decltype(length1))1;
+    const auto columnStart = decltype(length1)(1);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: same as above

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 6]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
 
-    auto column = new decltype(length1)[length1 + 1];
+    const auto column = new decltype(length1)[length1 + 1];
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: same as above

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 7]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
                 column[y - 1] + 1,
                 lastDiagonal + (string1[y - 1] == string2[x - 1] ? 0 : 1)
             };
-            column[y] = std::min(possibilities);
+            column[y] = min(possibilities);
             lastDiagonal = oldDiagonal;
         }
     }
-    auto result = column[length1];
+    const auto result = column[length1];
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: use of auto keyword

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 8]
```diff
@@ -525,40 +534,64 @@ uint64 CTS3Client::findChannelByNames(std::vector<std::string> details) {
...
+    const auto column = new decltype(length1)[length1 + 1];
     std::iota(column + columnStart, column + length1 + 1, columnStart);
 
     for (auto x = columnStart; x <= length2; x++) {
         column[0] = x;
         auto lastDiagonal = x - columnStart;
         for (auto y = columnStart; y <= length1; y++) {
-            auto oldDiagonal = column[y];
-            auto possibilities = {
+            const auto oldDiagonal = column[y];
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: use of auto keyword

**Review by TheMagnetar — CHANGES_REQUESTED** [review]
It looks really good. It is just those uses of auto keyword. I will update the C++ coding guidelines to reflect this.

**Comment by tbeswick96** [comment]
Changes done

**Comment by TheMagnetar** [comment]
Much appreciated!

**Review by TheMagnetar — APPROVED** [review]

**Pull request merged**